[JENKINS-46175] SimpleBuildStep/SimpleBuildWrapper without workspace (alternative approach)

Zastai opened the pull request:
See [JENKINS-46175](https://issues.jenkins-ci.org/browse/JENKINS-46175).

This is an alternative to https://github.com/jenkinsci/jenkins/pull/4784.
Instead of keeping the existing API, allowing some parameters to be null, this adds new methods to handle that case.
To avoid needing multiple new methods, this removes the separate choice for requiring a workspace and/or launcher.

### Proposed changelog entries

* JENKINS-46175: A `SimpleBuildStep` or `SimpleBuildWrapper` can now choose not to require a workspace context (working directory and launcher).

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed upgrade guidelines` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

Reference implementation:

- https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/123

### Desired reviewers

@jglick @oleg-nenashev @dwnusbaum

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
[JENKINS-46175] Add API to SimpleBuildStep

Two new methods are added:
- requiresWorkspace(), defaulted to returning true
- an overload of perform() that only takes a Run, EnvVars and listener
  - the "full-fat" perform() delegates to this one when requiresWorkspace() returns false

This enables steps to declare they do not need a workspace context.
BuildStepCompatibilityLayer honors this and won't try to get a workspace when one is not required.

Zastai committed:
[JENKINS-46175] Add API to SimpleBuildWrapper

These changes mirror those in SimpleBuildStep.

Disposer has the same changes applied except that the choice of requiring a workspace is made as part of the constructor, based on a wrapper instance.
Alternatively, it could choose this on its own, the way the step and wrapper do.

Zastai commented:
Setting as draft for now, just so the basic idea can be reviewed and compared to https://github.com/jenkinsci/jenkins/pull/4784

Zastai commented:
There seem to be some infra issues. The ATH step completed, reporting success, but then there are lots of stacktraces apparently caused by an AWS node going offline.

MarkEWaite commented:
Yes, sorry about the unreliability of the ci.jenkins.io agents.  We haven't yet been able to free enough time to investigate the root cause of the EC2 agents not remaining consistently connected to the ci.jenkins.io server.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -104,15 +104,51 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
             this.perform(run, workspace, launcher, listener);
         } else {
-            throw new AbstractMethodError();
+            throw new AbstractMethodError(Messages.SimpleBuildStep_NeedPerformMethodWithWorkspace());
```
This is for developers so should not be a localized message.

Zastai replied:
Figured as much; if this path is taken, I'll delocalize them.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -104,15 +104,51 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+    default void perform(@NonNull Run<?, ?> run, @NonNull EnvVars env, @NonNull TaskListener listener) throws InterruptedException, IOException {
+        // If this step requires a workspace, this is the wrong method to call.
+        if (this.requiresWorkspace()) {
+            throw new AbortException(Messages.SimpleBuildStep_WorkspaceContextNeeded());
```
ditto, and should be an `IllegalStateException` or similar

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -104,15 +104,51 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+            throw new AbortException(Messages.SimpleBuildStep_WorkspaceContextNeeded());
         }
+        // Otherwise, this method must have an implementation.
+        throw new AbstractMethodError(Messages.SimpleBuildStep_NeedPerformMethodWithoutWorkspace());
```
ditto

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -117,6 +156,32 @@ public void setDisposer(@NonNull Disposer disposer) {
+            this.requiresWorkspace = wrapper.requiresWorkspace();
+        }
+
+        private final boolean requiresWorkspace;
```
Invert the sense of the field since a deserialized `Disposer` from a Jenkins session prior to upgrade (for Pipeline builds) will have fields set to default values like `false` unless you implement `readResolve`.

Zastai replied:
Do these actually get persisted? I thought they were only serialized for sending to the agent for execution.

But sure, inverting the field's meaning is easy.
Or, I could just make this a method like on the other objects, avoiding the serialization issue entirely.

jglick replied:
`Disposer`s are _not_ sent to the agent. They are serialized in a Pipeline build, so that you can write for example

```groovy
xvnc {
  sh 'make world'
}
```

and have Jenkins restart in the middle of the `make` command but still know what to clean up later.

jglick replied:
> a method like on the other objects

As noted in the other PR, I would prefer to have a single flag for an implementor that covers both setup and teardown.

Zastai replied:
I know - but unfortunately neither the creation nor the execution of a `Disposer` is directly tied to a `SimpleBuildWrapper`.
Unless it would be acceptable to add a constructor to `Disposer` that takes a `SimpleBuildWrapper`; then it could set its internal flag from that.

jglick replied:
see https://github.com/jenkinsci/jenkins/pull/4829#discussion_r450575943

Zastai replied:
OK, will look at how to arrange it like that.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -126,7 +191,32 @@ public void setDisposer(@NonNull Disposer disposer) {
+                this.tearDown(build, listener);
+                return;
+            }
+            throw new AbstractMethodError(Messages.SimpleBuildWrapper_Disposer_NeedTearDownMethodWithWorkspace());
```
no i18n

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -126,7 +191,32 @@ public void setDisposer(@NonNull Disposer disposer) {
+        public void tearDown(Run<?,?> build, TaskListener listener) throws IOException, InterruptedException {
+            // If this callback requires a workspace, this is the wrong method to call.
+            if (this.requiresWorkspace) {
+                throw new AbortException(Messages.SimpleBuildWrapper_Disposer_WorkspaceContextNeeded());
```
ditto

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -126,7 +191,32 @@ public void setDisposer(@NonNull Disposer disposer) {
+                throw new AbortException(Messages.SimpleBuildWrapper_Disposer_WorkspaceContextNeeded());
+            }
+            // Otherwise, this method must have an implementation.
+            throw new AbstractMethodError(Messages.SimpleBuildWrapper_Disposer_NeedTearDownMethodWithoutWorkspace());
```
ditto

jglick commented on core/src/main/resources/jenkins/tasks/Messages.properties:
```diff
@@ -0,0 +1,9 @@
+SimpleBuildStep.NeedPerformMethodWithWorkspace=Unless a build step is marked as not requiring a workspace context, you must implement the overload of the perform() method that takes both a workspace and a launcher.
+SimpleBuildStep.NeedPerformMethodWithoutWorkspace=When a build step is marked as not requiring a workspace context, you must implement the overload of the perform() method that does not take a workspace or launcher.
+SimpleBuildStep.WorkspaceContextNeeded=This build step requires a workspace context, but none was provided.
+SimpleBuildWrapper.NeedSetUpMethodWithWorkspace=Unless a build wrapper is marked as not requiring a workspace context, you must implement the overload of the setUp() method that takes both a workspace and a launcher.
+SimpleBuildWrapper.NeedSetUpMethodWithoutWorkspace=When a build wrapper is marked as not requiring a workspace context, you must implement the overload of the setUp() method that does not take a workspace or launcher.
+SimpleBuildWrapper.WorkspaceContextNeeded=This build wrapper requires a workspace context, but none was provided.
+SimpleBuildWrapper_Disposer.NeedTearDownMethodWithWorkspace=Unless an end-of-wrapped-block callback is marked as not requiring a workspace context, you must implement the overload of the tearDown() method that takes both a workspace and a launcher.
+SimpleBuildWrapper_Disposer.NeedTearDownMethodWithoutWorkspace=When an end-of-wrapped-block callback is marked as not requiring a workspace context, you must implement the overload of the tearDown() method that does not take a workspace or launcher.
+SimpleBuildWrapper_Disposer.WorkspaceContextNeeded=This end-of-wrapped-block callback requires a workspace context, but none was provided.
```
ditto

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -75,7 +75,46 @@
+            this.setUp(context, build, listener, initialEnvironment);
+            return;
+        }
+        throw new AbstractMethodError(Messages.SimpleBuildWrapper_NeedSetUpMethodWithWorkspace());
```
ditto

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -75,7 +75,46 @@
+    public void setUp(Context context, Run<?,?> build, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException {
+        // If this wrapper requires a workspace, this is the wrong method to call.
+        if (this.requiresWorkspace()) {
+            throw new AbortException(Messages.SimpleBuildWrapper_WorkspaceContextNeeded());
```
ditto

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -75,7 +75,46 @@
+            throw new AbortException(Messages.SimpleBuildWrapper_WorkspaceContextNeeded());
+        }
+        // Otherwise, this method must have an implementation.
+        throw new AbstractMethodError(Messages.SimpleBuildWrapper_NeedSetUpMethodWithoutWorkspace());
```
ditto

jglick reviewed (commented):
Arguably simpler to understand, though a bit weaker since it does not permit a step to use a workspace context when one is provided while still operating when one is not.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -117,6 +156,32 @@ public void setDisposer(@NonNull Disposer disposer) {
+        /**
+         * Creates a new end-of-wrapped-block callback.
+         *
+         * @deprecated Use the constructor that takes a wrapper instance instead.
```
Should not be necessary. `Context` can be told whether its associated wrapper requests a workspace; then `setDisposer` can pass this bit of information along to the `Disposer`.

Zastai replied:
That is possible for non-pipeline cases only. When we're dealing with `AbstractBuild`, then the final overrides of `setUp()` and `preCheckout()` offer a place where I can tell the `Context` whether or not the wrapper needs a workspace context. Then I can just use a private `Boolean` field on both `Context` and `Disposer` to pass the flag across.

But for pipeline use, all methods involved are virtual and non-final, so there does not seem to be any place in code where I can guarantee interaction between the created `Context` and the `SimpleBuildWrapper` instance. The equivalent place in `CoreWrapperStep` would be the `doStart()` method. But then I would have to add public API to `Context` to flag the requirement there, which feels worse than simply allowing the `Dispatcher` to be told at construction time.

jglick replied:
> add public API to `Context` to flag the requirement there

Zastai replied:
Are there really no use cases supported other than `AbstractBuild` or pipeline? If not, then I suppose it's fine as long as `workflow-basic-steps` sets things up correctly.

jglick replied:
> Are there really no use cases supported other than `AbstractBuild` or pipeline?

No, that is it.

Zastai replied:
OK, I think I have something that works. Note that I think this does mean that when an old version of `workflow-basic-steps` is used with new core + a plugin with a wrapper marked as no-workspace-required, then any `Disposer` for that wrapper _will_ require that context and potentially fail (because the `Context.setWorkspaceRequirement()` won't have been called).

oleg-nenashev commented:
Personally I prefer this pull request to https://github.com/jenkinsci/jenkins/pull/4784 (here I respectfully disagree with @jglick, but it is rather a matter of the coding style and preferences than a functional difference). First thing is that I would want to merge one of the implementations before the LTS cutoff.

Zastai commented:
I’ll try to address the i18n remarks and generally get it to the same level of polish as #4784 tomorrow, so that either is as ready for merge as can be.

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai committed:
Remove i18n for developer messages

Zastai committed:
Adjust Disposer setup

This now uses a field with inverted meaning (for compatibility with
old serialized instances).
It also adds a constructor taking a boolean as well as a
requiresWorkspace() predicate method (for use by callers to determine
which of the tearDown() methods to call).
Extended JavaDoc for the tearDown methods.

Zastai committed:
Adjust exception use

Use IllegalStateException() instead of AbortException().
Tweaked and aligned JavaDoc.
Adjusted method order.

Zastai committed:
More Disposer changes

Its workspace requirement now tracks that of the wrapper via Context.

Zastai committed:
Finish a sentence in a comment.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -93,28 +93,76 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+     * @throws AbstractMethodError  if this step requires a workspace context and this method is not overridden
      * @throws InterruptedException if the step is interrupted
-     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @throws IOException          if something goes wrong; use {@link AbortException} for a polite error
```
I do not think we generally try to align Javadoc.

```suggestion
     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
```

Zastai replied:
It's something IntelliJ does; I can remove the indents manually if needed.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -64,18 +64,69 @@
+     * @return {@code true} if this wrapper requires a workspace context; {@code false} otherwise.
+     * @since TODO
+     */
+    boolean requiresWorkspace() {
```
```suggestion
    protected boolean requiresWorkspace() {
```

as you will see when you try to write a test for this

Zastai replied:
Right.

Zastai replied:
Actually, no. This needs to be `public` - it's the method you override to show you don't need a workspace context, which is information the caller needs to call the right `setUp` overload.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -97,6 +148,17 @@ public void env(String key, String value) {
+        /**
+         * Keeps track of whether or not the specified wrapper requires a workspace context (working directory and
+         * launcher).
+         * <p>
```
Should be either `@Restricted(Beta.class)` or otherwise make it clear that an implementation of `SimpleBuildWrapper` can ignore this.

Zastai replied:
Not sure I follow. This is on `Context`, so an implementation of `SimpleBuildWrapper` generally does not need to care anyway.
And I'm not sure why the restriction to Beta would make sense; for example, `workflow-basic-steps` needs to use this.

jglick replied:
> This is on `Context`, so an implementation of `SimpleBuildWrapper` generally does not need to care anyway.

Implementations would call other methods on `Context`.

> for example, `workflow-basic-steps` needs to use this

And that is intended to be the _only_ example. `Beta` would just be a way of reminding other plugins not to use it.

Not important, just one suggestion to help clarify that this is part of the implementation, not the normal API.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -117,16 +182,70 @@ public void setDisposer(@NonNull Disposer disposer) {
+        /**
+         * Determines whether or not this end-of-wrapped-block callback requires a workspace context (working
+         * directory and launcher).
+         * <p>
```
Similarly here.

Zastai replied:
Not sure why this would be restricted - the caller needs this info to call the right `tearDown` overload.

jglick replied:
Again, the only external caller should be in `workflow-basic-steps`, so this is a reminder that “regular” plugins should not touch it.

Zastai replied:
Oh, so `workflow-basic-steps` is already using a `useBeta` flag? Fair enough then.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -117,16 +182,70 @@ public void setDisposer(@NonNull Disposer disposer) {
      */
     public static abstract class Disposer implements Serializable {
+
+        private Boolean wrapperRequiresWorkspace;
```
```suggestion
        @CheckForNull
        private Boolean wrapperRequiresWorkspace;
```

Zastai replied:
Done

jglick reviewed (commented):
Generally seems fine. As previously noted, this approach suffers from the limitation that you cannot declare a step which _can_ accept a workspace but does not _require_ one. I cannot think of a case where this has cropped up in a real `Step` however, so it is probably unimportant.

Zastai commented:
> 
> 
> Generally seems fine. As previously noted, this approach suffers from the limitation that you cannot declare a step which _can_ accept a workspace but does not _require_ one. I cannot think of a case where this has cropped up in a real `Step` however, so it is probably unimportant.

Well, on the step/wrapper side you _can_. You override `requiresWorkspace` to return `false` and override both overloads of the relevant method (`perform`/`setUp`/`tearDown`).
What I'd need to do is to look at call sites and always call the overload that takes the workspace context when that context is available (if not overridden it forwards to the no-workspace version anyway). Should not be too difficult.

Zastai committed:
Correct method accessibility.

While `SimpleBuildStep` is an interface, `SimpleBuildWrapper` is a
class, so needs an explicit `public`.

Zastai committed:
Add nullability annotations

Zastai committed:
Undo JavaDoc param/throws alignment

Zastai committed:
Enable optional-but-handled workspace context

This adjusts the call sites to always call the with-workspace-context
methods when such a context is available, regardless of whether it is
required.

Zastai committed:
Correct JavaDoc

Zastai committed:
Mark internal API

This adds `@Restricted(Beta.class)` as suggested, to indicate methods
that are part of an internal-only API (essentially they should be
called by the workflow-basic-steps plugin only).
(Does feel like there should be a better tag than "BETA" to indicate
this.)

MRamonLeon commented:
I prefer this approach because it enforces the parameters and methods called.

timja commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -64,18 +66,70 @@
+     * @return {@code true} if this wrapper requires a workspace context; {@code false} otherwise.
+     * @since TODO
+     */
+    @Restricted(Beta.class) // to indicate it is to be called by Jenkins internals only; not part of the normal API
```
this comment doesn't seem right? beta just means that it's a new API and it may change in some way? To me anyway

jglick replied:
https://github.com/kohsuke/access-modifier/issues/25

Zastai replied:
It's also why I added the comment, to make it clear its not "regular" `Beta`.

jglick replied:
```suggestion
```

Wrong here.

timja reviewed (commented):
slight preference to this approach

Zastai commented:
Given this version seems to have the most approval so far, I've un-drafted this one, and converted the other one to draft.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -54,6 +54,7 @@
 import jenkins.model.RunAction2;
 import jenkins.model.TransientActionFactory;
 import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
```
```suggestion
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -159,6 +160,7 @@ default void perform(@NonNull Run<?, ?> run, @NonNull EnvVars env, @NonNull Task
      * @return {@code true} if this step requires a workspace context; {@code false} otherwise.
      * @since TODO
      */
+    @Restricted(Beta.class) // to indicate it is to be called by Jenkins internals only; not part of the normal API
```
```suggestion
```

No, _this_ is key to the public API for plugins implementing build steps.

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai committed:
Remove Beta restriction from requiresWorkspace

Zastai committed:
Remove Beta restriction from requiresWorkspace

Zastai committed:
Add back a Beta restriction

The one the Disposer is not for implementers to override.
I need more coffee.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -93,28 +93,76 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+     * Run this step, without a workspace context.
+     * <p>
+     * This method <strong>must</strong> be overridden when this step does not require a workspace context, and
+     * <strong>must not</strong> be called when such a context <em>is</em> required.
```
```suggestion
     * will not be called when such a context <em>is</em> required.
```

(write from the PoV of an implementor)

Zastai replied:
OK. I generally always write docs for the user of the method. But this change makes sense, and kind of works for the user/caller too.

jglick replied:
My point is that the Javadoc is mostly going to read by plugin implementors of build steps/wrappers, and this emphasis was only relevant for the two call sites (in core and `workflow-basic-steps`).

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -64,18 +66,69 @@
+     * Called when a segment of a build is started that is to be enhanced with this wrapper.
+     * <p>
+     * This method <strong>must</strong> be overridden when this wrapper does not require a workspace context, and
+     * <strong>must not</strong> be called when such a context <em>is</em> required.
```
```suggestion
     * will not be called when such a context <em>is</em> required.
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -143,6 +268,7 @@ protected boolean runPreCheckout() {
             return new Environment() {};
         } else {
             final Context c = new Context();
+            c.setWorkspaceRequirement(this);
```
Purely stylistic, but to make all this logic easier to follow:

* remove `setWorkspaceRequirement`
* deprecate the implicit no-arg `Context` constructor (by first declaring it explicitly)
* introduce a private (or package-scope) constructor taking a `boolean requiresWorkspace` argument
* call that from a new “beta” method `SimpleBuildWrapper.createContext()`
* then you can

```suggestion
            final Context c = createContext();
```

Zastai replied:
OK

Zastai replied:
Done

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -172,7 +299,7 @@ protected boolean runPreCheckout() {
         @Override public boolean tearDown(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
             if (c.disposer != null) {
-                c.disposer.tearDown(build, build.getWorkspace(), launcher, listener);
+                c.disposer.tearDown(build, build.getWorkspace(), this.launcher, listener);
```
Note: not actually necessary. Not sure if there is some style guide saying to use `this.` when not necessary to disambiguate but merely as a reminder that the reference is to a field.

Zastai replied:
I tend towards always qualifiying with `this` (or the declaring class for static items) because I find it really helps when reading new codebases. I can drop this diff, no problem.

jglick replied:
Whatever you find more legible is fine.

Zastai replied:
Shrug. In this particular case, the class is so small it hardly matters either way, so I'll leave the change out of the PR.

Zastai commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -93,28 +93,76 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+     * @throws AbstractMethodError if this step requires a workspace context and neither this method nor {@link #perform(Run, FilePath, Launcher, TaskListener)} is overridden
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
      * @since TODO
```
Side note: I thought these were supposed to be updated as part of the release process? This particular method was added in 2.241 I think.

Zastai replied:
(This pointed to the `@since TODO` but apparently does not get updates by GH when more commits are added)

jglick replied:
Not sure, @daniel-beck?

timja replied:
Contributors update them every now and then, it’s not automated

Zastai committed:
Tweak JavaDoc as suggested by @jglick

This also removes some unnecessary diffs (one `this.` qualifier), and
one automated import ordering.

Zastai committed:
Adjust wrapper context creation

As suggested by @jglick; a Context is now canonically created via
SimpleBuildWrapper.createContext(), which will tell the context
whether or not there is a workspace requirement (which is then
passed on to any Disposer as before).

jglick reviewed (approved):
This looks clearer to me indeed.

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai commented:
You know, it's pretty disappointing that the CI for a CI product is so broken... I'm trying to generate an incremental so that I can update the PR for worflow-basic-steps, but it seems to be failing all over the place (`HTTP Error 500.30 - ANCM In-Process Start Failure` this time).

jglick commented:
Hmm, yeah, could you file an `INFRA` ticket for that please?

timja commented:
I did have a look in the logs and couldn’t see anything around that time, will trigger a rebuild and see if it’s still happening

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai commented:
Showing as 1 failed test now, but that looks like an unrelated broken test, based on the
```
expected:<...:true,"jvm-vendor":"[N/A]","jvm-name":"OpenJD...> but was:<...:true,"jvm-vendor":"[Oracle Corporation]","jvm-name":"OpenJD...>
```

jglick commented:
Already fixed: #4879

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai commented:
@oleg-nenashev I take it this has now missed the cut for the next LTS?

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai commented:
Great - more unrelated build failures.
```
Aug 23, 2020 2:02:10 PM hudson.model.UpdateCenter verifyChecksums
INFO: Attempt to verify a downloaded file (example) using SHA-512 or SHA-256 failed since your configured update site does not provide either of those checksums. Falling back to SHA-1.
Aug 23, 2020 2:02:10 PM hudson.model.UpdateCenter verifyChecksums
INFO: Attempt to verify a downloaded file (example) using SHA-512 or SHA-256 failed since your configured update site does not provide either of those checksums. Falling back to SHA-1
...[truncated 456 bytes]...
Falling back to SHA-1.
Aug 23, 2020 2:02:10 PM hudson.model.UpdateCenter verifyChecksums
WARNING: Attempt to verify a downloaded file (example) using SHA-512 failed since it could not be computed. Falling back to weaker algorithms. Update your JRE.
Aug 23, 2020 2:02:10 PM hudson.model.UpdateCenter verifyChecksums
INFO: Attempt to verify a downloaded file (example) using SHA-512 or SHA-256 failed since your configured update site does not provide either of those checksums. Falling back to SHA-1.
```
on both JDK8 and JDK11.

MarkEWaite commented:
I've started that build again just in case the issue is transient.  

I've also started a rebuild of two other builds, [PR-4912](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4912/2/) and [PR-4913](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4913/2/) in case the failures are related to recent improvements in the Jenkins update center.  If those two PR builds show the same failure, then we'll likely need to request help from @daniel-beck and @timja .

timja commented:
> I've started that build again just in case the issue is transient.
> 
> I've also started a rebuild of two other builds, [PR-4912](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4912/2/) and [PR-4913](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4913/2/) in case the failures are related to recent improvements in the Jenkins update center. If those two PR builds show the same failure, then we'll likely need to request help from @daniel-beck and @timja .

The tests pass locally for me

Zastai commented:
Same failure again. The "please upgrade your JRE" for the SHA-512 processing bothers me a little. Either the message is bogus, or there's an issue with the JDK/JRE in use for the build (which seems unlikely).

MarkEWaite commented:
> Same failure again. The "please upgrade your JRE" for the SHA-512 processing bothers me a little. Either the message is bogus, or there's an issue with the JDK/JRE in use for the build (which seems unlikely).

Same failure happens in #4912 [build](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4912/2/) and #4913 [build](https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-4913/2/testReport/).  I suspect that we'll need help from @daniel-beck or @timja to understand the error in the test or the error in the most recent changes to the update center.  Those tests were passing a few days ago and now are failing consistently.

MarkEWaite commented:
In the failure case, the maven and Java versions are reported as:

```
13:16:23  Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
13:16:23  Maven home: /usr/share/maven
13:16:23  Java version: 1.8.0_265, vendor: Oracle Corporation, runtime: /usr/local/openjdk-8/jre
```

In the success case of the preceding build from 3 days ago, the maven and Java versions are:

```
12:07:55  Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
12:07:55  Maven home: /usr/share/maven
12:07:55  Java version: 1.8.0_265, vendor: Oracle Corporation, runtime: /usr/local/openjdk-8/jre
```

MarkEWaite commented:
@daniel-beck and @timja I see the same failure reported when I use Java 8u262 on Ubuntu 18.04 to run:

```
$ mvn clean -Dtest=UpdateCenterTest test
```

Test fails and the info mentions that SHA-512 and SHA-256 are not found.  Maybe the file names have changed for the SHA-xxx files?

timja commented:
> @daniel-beck and @timja I see the same failure reported when I use Java 8u262 on Ubuntu 18.04 to run:
> 
> ```
> $ mvn clean -Dtest=UpdateCenterTest test
> ```
> 
> Test fails and the info mentions that SHA-512 and SHA-256 are not found. Maybe the file names have changed for the SHA-xxx files?

ffs, there's two UpdateCenterTest's I ran the other one.
https://github.com/jenkinsci/jenkins/pull/4915

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai committed:
Merge branch 'master' into JENKINS-46175B

Zastai commented:
This seems to be languishing; merged in master because it had been a while.

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
